src/mesh.f90

```diff
@@ -0,0 +1,325 @@
+  type :: mesh_t
+    integer, dimension(3), private :: dims_global ! global number of vertices in each direction
+
+    integer, dimension(3), private :: cdims_padded ! local domain size including padding (cartesian structure)
+    integer, dimension(3), private :: cdims ! local domain size without padding (cartesian structure)
+    logical, dimension(3), private :: periodic_dir ! Whether or not a direction has a periodic BC
+    integer, private :: sz
+    class(geo_t), allocatable :: geo ! object containing geometry information
+    class(parallel_t), allocatable :: par ! object containing parallel domain decomposition information
+  contains
+    procedure :: get_SZ
+
+    procedure :: get_n_groups_dir
+    procedure :: get_n_groups_phi
+    generic :: get_n_groups => get_n_groups_dir, get_n_groups_phi
+
+    procedure :: get_dims_dir
+    procedure :: get_dims_phi
+    procedure :: get_dims_phi_dataloc
+    generic :: get_dims => get_dims_dir, get_dims_phi, get_dims_phi_dataloc
+
+    procedure :: get_n_dir 
+    procedure :: get_n_phi
+    generic :: get_n => get_n_dir, get_n_phi
+
+    procedure :: get_padded_dims_phi
+    procedure :: get_padded_dims_dir
+    generic :: get_padded_dims => get_padded_dims_dir, get_padded_dims_phi
+  
+    procedure :: get_location
+  end type mesh_t
+
+  interface mesh_t
+    module procedure mesh_init
+  end interface mesh_t
+
+  contains 
+
+  function mesh_init(dims_global, nproc_dir, L_global, sz) result(mesh)
+    !! Completely initialise the mesh object. 
+    !! Upon initialisation the mesh object can be read-only and shouldn't be edited
+    !! Takes as argument global information about the mesh like its length, number of cells and decomposition in each direction
+    integer, dimension(3), intent(in) :: dims_global
+    integer, dimension(3), intent(in) :: nproc_dir ! Number of proc in each direction
+    real(dp), dimension(3), intent(in) :: L_global
+    integer, intent(in) :: sz
+    type(mesh_t) :: mesh
+
+    integer :: nx_padded, ny_padded, nz_padded
+    integer :: nx, ny, nz 
+    integer :: ierr
+
+    allocate(mesh%geo)
+    allocate(mesh%par)
+    mesh%dims_global(:) = dims_global
+    ! Geometry
+    mesh%geo%L = L_global
+    mesh%geo%d = mesh%geo%L(:) / mesh%dims_global(:)
+
+    ! Parallel domain decomposition
+    mesh%par%nproc_dir(:) = nproc_dir
+    mesh%par%nproc = product(nproc_dir(:))
+    call MPI_Comm_rank(MPI_COMM_WORLD, mesh%par%nrank, ierr)
+    call MPI_Comm_size(MPI_COMM_WORLD, mesh%par%nproc, ierr)
+    call domain_decomposition(mesh)
+
+    ! Local mesh dimensions
+    nx = mesh%dims_global(1)/mesh%par%nproc_dir(1)
+    ny = mesh%dims_global(2)/mesh%par%nproc_dir(2)
+    nz = mesh%dims_global(3)/mesh%par%nproc_dir(3)
+
+    ! Apply padding based on sz
+    nx_padded = nx - 1 + mod(-(nx - 1), sz) + sz
+    ny_padded = ny - 1 + mod(-(ny - 1), sz) + sz
+    ! Current reorder functions do not require a padding in z-direction.
+    nz_padded = nz
```

Comment:
Padding is actually related to the allocator more than anything else but now its getting sorted inside the mesh object. It can be better if it is handled by the allocator instead.